### PR TITLE
Fix issue #4166

### PR DESCRIPTION
**Pull request by AllenX2018**
Fix issue #4166 : Add cache-aside pattern

I only add cache-aside pattern in this pr.
If the implement of the cache-pattern in this pr is ok, I will try implement of other cache pattern, such as read-through, write-through...

### CODE REVIEW

File: client/cache/cache_pattern.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	if len(keysErr) == 0 {
+		return vals, nil
+	}
+
+	return vals, fmt.Errorf(strings.Join(keysErr, "; "))
+}
+
+func (c *CacheAside) Put(ctx context.Context, key string, val interface{}, timeout time.Duration) error {
```
Comment (jianzhiyao) [inline client/cache/cache_pattern.go:64]: this operations shold be locked?

Comment (AllenX2018) [reply client/cache/cache_pattern.go:64]: > this operations should be locked?

I think the lock will be implemented by `cache`, `reader` and `writer` passed to `NewCacheAside`.
For example, if we pass `MemoryCache` to `NewCacheAside`,  `CacheAside.Put` will use lock in `MemoryCache.Delete` while deleting cache.

File: client/cache/cache_pattern.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+	val, err := c.reader.Query(key)
+	if err != nil {
+		return nil, err
+	}
+
+	return val, c.Cache.Put(ctx, key, val, c.lifeSpan)
+}
+
+func (c *CacheAside) GetMulti(ctx context.Context, keys []string) ([]interface{}, error) {
```
Comment (flycash) [inline client/cache/cache_pattern.go:45]: I think add a new method named "QueryBatch" into `ValueReader` would be better. Users always expect that the `CacheAside` work the same way as `Cache`.

For example, users create the `CacheAside` instance by passing a redis Cache instance which read all values  from redis server in one command `MGET`.

But this method change the behavior.

File: client/cache/cache_pattern.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+// CacheAside is a decorator of Cache, it implements "cache aside" pattern
+type CacheAside struct {
+	Cache
+	reader   ValueReader
+	writer   ValueWriter
+	lifeSpan time.Duration
+}
+
+func NewCacheAside(cache Cache, reader ValueReader, writer ValueWriter, lifeSpan time.Duration) *CacheAside {
```
Comment (flycash) [inline client/cache/cache_pattern.go:28]: I think both reader, writer and lifeSpan are optional.

For example, users only want `CacheAside` read data from DB if the cache didn't have the specific key. But users update values by ORM directly:
```go

func demo() {
   aside.NewCacheAside(redisCache, ormReader)
   aside.Get(ctx, key1)
   orm.update(key1, xxx)
}

```

File: client/cache/cache_pattern.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+func (c *CacheAside) Put(ctx context.Context, key string, val interface{}, timeout time.Duration) error {
+	if err := c.writer.Update(key, val); err != nil {
+		return err
+	}
+
+	if ok, _ := c.IsExist(ctx, key); !ok {
+		return nil
+	}
+
+	return c.Cache.Delete(ctx, key)
```
Comment (flycash) [inline client/cache/cache_pattern.go:73]: we don't need to check whether exist, just delete it.

File: client/cache/cache_pattern.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+	if err != nil {
+		return err
+	}
+
+	if err := c.writer.Update(key, updateVal); err != nil {
+		return err
+	}
+
+	return c.Cache.Delete(ctx, key)
+}
```
Comment (flycash) [inline client/cache/cache_pattern.go:110]: Both two methods have concurrency issue.

**Review by flycash — CHANGES_REQUESTED** [review]
Please resolve the concurrency problem

File: client/cache/cache_pattern.go (hunk 3 above)
Comment (AllenX2018) [reply client/cache/cache_pattern.go:28]: > I think both reader, writer and lifeSpan are optional.
> 
> For example, users only want `CacheAside` read data from DB if the cache didn't have the specific key. But users update values by ORM directly:
> 
> ```go
> func demo() {
>    aside.NewCacheAside(redisCache, ormReader)
>    aside.Get(ctx, key1)
>    orm.update(key1, xxx)
> }
> ```

I think `CacheAside` encapsulates operation of data source and cache for user. If the user implements the data source operation by himself, I think it loses the meaning of `CacheAside`.

**Comment by AllenX2018** [comment]
> Please resolve the concurrency problem

OK. I will fix and commit after all problems solved.

**Comment by jianzhiyao** [comment]
After discussion with @AllenX2018 , close that pr for now.

**Pull request closed without merging**